pkg/s3select/select.go

```diff
@@ -276,91 +297,111 @@ func (reader *Input) processSelectReq(reqColNames []string, alias string, whereC
+		}
+
+	case *JSONInput:
+		// JSON doesnt have columns so no cleaning
 	}
+
 	return nil
 }
 
 // processColNameIndex is the function which creates the row for an index based
 // query.
-func (reader *Input) processColNameIndex(record []string, reqColNames []string, columns []string) (string, error) {
+func processColNameIndex(record string, reqColNames []string, columns []string, inputType SelectQuery) (string, error) {
 	myRow := ""
 	for i := 0; i < len(reqColNames); i++ {
 		// COALESCE AND NULLIF do not support index based access.
 		if reqColNames[0] == "0" {
 			return "", ErrInvalidColumnIndex
 		}
-		// Subtract 1 because AWS Indexing is not 0 based, it starts at 1.
 		mytempindex, err := strconv.Atoi(reqColNames[i])
-		mytempindex = mytempindex - 1
 		if mytempindex > len(columns) {
 			return "", ErrInvalidColumnIndex
 		}
-		myRow = writeRow(myRow, record[mytempindex], reader.options.OutputFieldDelimiter, len(reqColNames))
+		myRow = writeRow(myRow, jsonValue(string("_"+strconv.Itoa(mytempindex-1)), record), inputType.getOutPutFieldDelimitter(), len(reqColNames))
 		if err != nil {
 			return "", ErrMissingHeaders
 		}
 	}
 	if len(myRow) > 1000000 {
 		return "", ErrOverMaxRecordSize
 	}
-	if strings.Count(myRow, reader.options.OutputFieldDelimiter) != len(reqColNames)-1 {
-		myRow = qualityCheck(myRow, len(reqColNames)-1-strings.Count(myRow, reader.options.OutputFieldDelimiter), reader.options.OutputFieldDelimiter)
+	if strings.Count(myRow, inputType.getOutPutFieldDelimitter()) != len(reqColNames)-1 {
+		myRow = qualityCheck(myRow, len(reqColNames)-1-strings.Count(myRow, inputType.getOutPutFieldDelimitter()), inputType.getOutPutFieldDelimitter())
 	}
 	return myRow, nil
 }
 
 // processColNameLiteral is the function which creates the row for an name based
 // query.
-func (reader *Input) processColNameLiteral(record []string, reqColNames []string, columns []string, columnsMap map[string]int, myFunc *SelectFuncs) (string, error) {
+func processColNameLiteral(record string, reqColNames []string, myFunc *SelectFuncs, inputType SelectQuery) (string, error) {
 	myRow := ""
 	for i := 0; i < len(reqColNames); i++ {
 		// this is the case to deal with COALESCE.
 		if reqColNames[i] == "" && isValidFunc(myFunc.index, i) {
-			myVal := evaluateFuncExpr(myFunc.funcExpr[i], "", record, columnsMap)
-			myRow = writeRow(myRow, myVal, reader.options.OutputFieldDelimiter, len(reqColNames))
+			myVal := evaluateFuncExpr(myFunc.funcExpr[i], "", record)
+			myRow = writeRow(myRow, myVal, inputType.getOutPutFieldDelimitter(), len(reqColNames))
 			continue
 		}
-		myTempIndex, notFound := columnsMap[trimQuotes(reqColNames[i])]
-		if !notFound {
-			return "", ErrMissingHeaders
-		}
-		myRow = writeRow(myRow, record[myTempIndex], reader.options.OutputFieldDelimiter, len(reqColNames))
+		myRow = writeRow(myRow, jsonValue(reqColNames[i], record), inputType.getOutPutFieldDelimitter(), len(reqColNames))
 	}
 	if len(myRow) > 1000000 {
```

Comment:
Can we add a constant for this number somewhere?

Reply:
Is this more like, we return a max of 1 million records? Is it the same restriction with AWS?

Reply:
>Is this more like, we return a max of 1 million records? Is it the same restriction with AWS? No

Reply:
This should be in a constant.